### ISSUE TITLE
App freeze when stopping watcher and session soon after watcher creation.

### ISSUE THREAD

**Issue by Tolo789** [Azure/azure-spatial-anchors-samples#339]
**Description**
Hi,
In the Unity app I am working on I need to be able to start and stop the search for anchors at a moment notice and I systematically have a freeze of my iPhone when the stop happens within ~6s after having created the watcher.
I've tried to reproduce the same behaviour by tweaking your samples but it seems that the issue only happens on my project, however I've checked every step from the session creation to the session stop and we do everything alike..

Unluckily I cannot share my app whole code and the simpler app demo that I'm trying to create don't have the bug, but I was hoping that maybe you could give me some pointers on how I may be able to fix the issue anyways?

I also wanted to point out that I've managed to somehow avoid the freeze by switching to a background thread before stopping the watcher/session. However the operation still takes ~6s to perform, and thus have to await that it finishes before being able to start a new search.

**Steps to reproduce the issue**
These steps work fine in the Unity's ASA sample apps, it's just on my custom app that I have the issue.

- Create `Session` and `AnchorLocateCriteria`
```
_anchorManager = gameObject.AddComponent<SpatialAnchorManager>();
await _anchorManager.CreateSessionAsync();
_anchorManager.Session.Diagnostics.ImagesEnabled = true;
_anchorManager.Session.Diagnostics.LogLevel = SessionLogLevel.Warning;
_anchorManager.Session.Diagnostics.MaxDiskSizeInMB = 50;
_anchorManager.Session.TelemetryEnabled = false;

_anchorLocateCriteria = new AnchorLocateCriteria();
_anchorLocateCriteria.RequestedCategories = AnchorDataCategory.Spatial;
_anchorLocateCriteria.BypassCache = true;
_anchorLocateCriteria.Strategy = LocateStrategy.AnyStrategy;
```
- Start search
```
if (_watcher != null)
{
    return;
}
await _anchorManager.StartSessionAsync();
_anchorLocateCriteria.Identifiers = new string[]{Guid.NewGuid().ToString()};
_watcher = _spatialAnchorManager.Session.CreateWatcher(_anchorLocateCriteria);
```
- Stop search (within ~6s after having started it, otherwise no freeze happens)
```
if (_watcher != null)
{
    _watcher.Stop();
    _watcher = null;
    _spatialAnchorManager.StopSession();
}
```

**Expected behavior**
Stopping the search right after starting it shouldn't take ~6s (even on a background thread).

**Development information (please complete the following information)**
 - OS: macOS
 - Development Platform\Version: Xcode 13.1, Unity 2021.1.27f1
 - Azure Spatial Anchors SDK version: `2.11.0`

**AR Device information (please complete the following information):**
 - Device: iPhone SE 2nd gen (bug was seen also on other iPhones and iPods but I do not have the exact model)
 - OS: iOS 15.1

**Comment by Varnauld** [comment]
Thank you for the thorough outline! Appreciate the feedback

Tracking n# [38271124] - pending assignment. Will update this thread as soon as one of our team members is assigned,

**Comment by the-mad** [comment]
Hello there 👋 
I experienced something "similar": when I stop an ASA session when **an requested anchor has not been found**, then the app crashes.
Not sure if it's related, but I though I'd share it, just in case it could help !

**Comment by Varnauld** [comment]
Thank you @the-mad for this added information - apologies for the delay, this topic is still pending assignment but I hope to provide an update this week,

**Comment by darax** [comment]
@Tolo789 
In the start search block you have above is this in one of the functions unity calls each frame, like Update or LateUpdate? if so, one possible issue is reentrance:

if (_watcher != null)
{
    return;
}
await _anchorManager.StartSessionAsync(); // While this is awaiting the _watcher will remain null, and this could be called again
_anchorLocateCriteria.Identifiers = new string[]{Guid.NewGuid().ToString()};
_watcher = _spatialAnchorManager.Session.CreateWatcher(_anchorLocateCriteria);

**Comment by darax** [comment]
@the-mad  your issue sounds different. It might be better to file a separate one.

**Comment by Tolo789** [comment]
Hi @darax,
My start/stop search methods are triggered with an UI button by the user that behaves as a toggle, so I have the freeze whenever the button is clicked twice within 5s.

Thanks for the `AutoResetEvent` idea, personally I was using a semaphore and CancellationToken logic to know if my work in the background thread was finished or not and for now it was enough.

**Comment by the-mad** [comment]
> @the-mad your issue sounds different. It might be better to file a separate one.

Hi 👋 
I made a separated issue here https://github.com/Azure/azure-spatial-anchors-samples/issues/346

Thx in advance !

**Comment by darax** [comment]
@Tolo789 
I see, the watcher == null is your sentinal value to know if the StartSessionAsync has not yet completed.
It might be interesting to set the [SessionLogLevel](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.spatialanchors.sessionloglevel?view=spatialanchors-dotnet) to 'All' and see if maybe we're hitting some HTTP (or other) timeout early in our processing.

**Comment by Tolo789** [comment]
Hey @darax , the toggle basically works like this:
```
private bool _startSearch = false;
private bool _blockToggle = false;

// Method called whenever the UI button is clicked.
public async UniTask ToggleSearch()
{
    if (_blockToggle)
        return;
    _blockToggle = true;

    try
    {
        _startSearch = !_startSearch;
        if (_startSearch)
            await this.StartSearch();
        else
            this.StopSearch();
    }
    finally
    {
        _blockToggle = false;
    }
}
```

What I had to do to solve my issue was to make StopSearch method async and use background thread like this:
```
private async UniTask StopAnchorSearchAsync()
{
    try
    {
        await UniTask.SwitchToThreadPool();
        if (_watcher != null)
        {
            _watcher.Stop();
            _watcher = null;
            _spatialAnchorManager.StopSession();
        }
    }
    finally
    {
        // Ensure to always return to main thread before exiting this method
        await UniTask.SwitchToMainThread();
    }
}
```

As you may have noticed we use `UniTask` instead of `Task` for our project, but I don't think it should compromise the usage of ASA?

Anyways will try to use `SessionLogLevel.All` to see if there is anything new in the logs

**Comment by Tolo789** [comment]
I've just run some tests with the `SessionLogLevel.All` and changing StopSearch method so that it doesn't use background thread hack, I confirm that I still have the freeze but no interesting log showed up :(

**Comment by XiangyuL-Microsoft** [comment]
@Tolo789 

We just released a new version of SDK [2.13.0](https://github.com/Azure/azure-spatial-anchors-samples/releases/tag/v2.13.0).
Could you try this new version of SDK to see if the problem still persists please?
Thank you very much.

**Comment by Tolo789** [comment]
Hi all, thank you for coming back to see if I still have the issue. Unluckily we dropped our usage of Azure Spatial Anchor so I can no longer test this out for you, feel free to close this ticket!
